R/renv.R

```diff
@@ -308,3 +308,56 @@ renv_switch_r_version <- function(version = NULL
 
   return(invisible(TRUE))
 }
+
+
+#' Downgrade an renv project to a specific RSPM snapshot
+#'
+#' @description
+#'   This functions aims to be used within a "snapshot-centered project
+#'   workflow" and can be used to downgrade all packages to an RSPM snapshot listed in
+#'   `renv.lock`.
+#'
+#'   While the main purpose is downgrade packages which exist in a higher
+#'   version, this function can also be used to restore a clean state of the
+#'   project library outside of a downgrade scenario.
+#'
+#'   Under the hood, it records all packages installed in the `renv` project
+#'   library and restores these with the RSPM snapshot found in Line 7 of
+#'   `renv.lock`.
+#' @return Called for its side-effect.
+#' @export
+#' @seealso renv_switch_r_version
+#'
+#' @examples
+#' \dontrun{
+#' renv_downgrade()
+#' }
+renv_downgrade <- function() {
+  requireNamespace("renv", quietly = TRUE)
+
+  # check if renv.lock exists
+  checkmate::assert_file("renv.lock", )
+
+  # to ensure repos is set correctly
+  source(".Rprofile")
+
+  # list all packages from renv library (should always be the renv lib
+  # if renv.lock exists)
+  installed_pkgs <- unname(utils::installed.packages(
+    lib.loc = .libPaths()[1]
+  )[, "Package"])
+
+  snapshot_date <- stringr::str_extract(
+    readLines("renv.lock", n = 7)[7],
+    "[0-9]{4}-[0-9]{2}-[0-9]{2}"
+  )
```

Comment:
Use `getOption("repos")` to get the snapshot date?

Reply:
I thought about it at first but it the user has not yet restarted the session, `getOption("repos")` will be wrong. I thought about restarting the session before taking action but this is not so trivial and `rstudioapi::restart_session()` is not fail-safe.

Hence I thought its safer to extract from what is found in `renv.lock`?

Reply:
If the user hasn't restarted the session, what package versions will `renv::install()` install?

Reply:
The ones from the snapshot found in `renv.lock`, see the `source(".Rprofile")` call within the function which updates the repos option.